src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs

```diff
@@ -0,0 +1,426 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Runtime.Serialization;
+using Akka.Actor;
+using Akka.Event;
+using Akka.Streams.Kafka.Extensions;
+using Akka.Streams.Kafka.Helpers;
+using Akka.Streams.Kafka.Settings;
+using Akka.Streams.Kafka.Stages.Consumers.Exceptions;
+using Akka.Util.Internal;
+using Confluent.Kafka;
+using Newtonsoft.Json;
+
+namespace Akka.Streams.Kafka.Stages.Consumers.Actors
+{
+    internal class KafkaConsumerActor<K, V> : ActorBase
+    {
+        private readonly IActorRef _owner;
+        private readonly ConsumerSettings<K, V> _settings;
+        private readonly IConsumerEventHandler _consumerEventHandler;
+        
+        private ICancelable _poolCancellation;
+        private Internal.Poll<K, V> _pollMessage;
+        private Internal.Poll<K, V> _delayedPollMessage;
+        
+        private IImmutableDictionary<IActorRef, KafkaConsumerActorMetadata.Internal.RequestMessages> _requests 
+            = ImmutableDictionary<IActorRef, KafkaConsumerActorMetadata.Internal.RequestMessages>.Empty;
+        private IImmutableSet<IActorRef> _requestors = ImmutableHashSet<IActorRef>.Empty;
+        private IConsumer<K, V> _consumer;
+        private readonly ILoggingAdapter _log;
+        private bool _stopInProgress = false;
+        private bool _delayedPoolInFlight = false;
+
+        /// <summary>
+        /// While `true`, committing is delayed.
+        /// Changed by `onPartitionsRevoked` and `onPartitionsAssigned` callbacks
+        /// </summary>
+        private bool _rebalanceInProgress = false;
+        /// <summary>
+        /// Keeps commit offsets during rebalances for later commit.
+        /// </summary>
+        private IImmutableSet<TopicPartitionOffset> _rebalanceCommitStash = ImmutableHashSet<TopicPartitionOffset>.Empty;
+        /// <summary>
+        /// Keeps commit senders that need a reply once stashed commits are made.
+        /// </summary>
+        private IImmutableList<IActorRef> _rebalanceCommitSenders = new ImmutableArray<IActorRef>();
+
+        public KafkaConsumerActor(IActorRef owner, ConsumerSettings<K, V> settings, IConsumerEventHandler consumerEventHandler)
+        {
+            _owner = owner;
+            _settings = settings;
+            _consumerEventHandler = consumerEventHandler;
+            
+            _pollMessage = new Internal.Poll<K, V>(this, periodic: true);
+            _delayedPollMessage = new Internal.Poll<K, V>(this, periodic: false);
+            _log = Context.GetLogger();
+        }
+
+        protected override bool Receive(object message)
+        {
+            switch (message)
+            {
+                case KafkaConsumerActorMetadata.Internal.Assign assign:
+                {
+                    _log.Debug($"Got assign request from {Sender}");
```

Comment:
One performance nitpick - better off using the old school string formatting syntax than string interpolation, because due to the way Akka.NET's logging is designed we can actually defer allocation of strings until we're ready to print the log statement out to console using the old technique. If the LogLevel is set at INFO, which is is by default, this saves us an unnecessary allocation each time one of these `Assign` messages is processed.

Not so with the new one. I'm not sure how "high traffic" this section of actor code could be, but that's something to bear in mind. It can really add up in busy ActorSystems.

Reply:
Never thought about that... This needs to be reviewed in a whole code base of this project, because I am using string interpolation whenever possible... Created separate issue for this #49 to do later